### PR TITLE
feat: add `ignore` option

### PR DESCRIPTION
**Pull request by btea**

### CODE REVIEW

**Comment by sindresorhus** [comment]
AI review. Haven't double-checked it:

---

### ❌ `path.pop()` is missing in some branches

* In `deepSortArray`:

  ```js
  if (Array.isArray(item)) {
  	return deepSortArray(item); // ❌ Missing path.pop()
  }

  if (isPlainObject(item)) {
  	return _sortKeys(item); // ❌ Missing path.pop()
  }
  ```

  → Fix: Add `path.pop()` before returning in both branches.

* In `_sortKeys`, `path.pop()` is only called after setting `newValue`, but the early `continue` already does a `path.pop()`, so that’s fine.

---

### ❌ `index` used as a key

```js
key: index
```

* This violates the declared type `key: string` in your `ignore` function.

→ Fix: Cast: `key: String(index)`

---

### ❌ `path` mutation is shared across recursive calls

* `path` is mutated in-place (`push`/`pop`) but reused across all recursive branches.
* This works due to strict discipline but is fragile and breaks if an `await` or async change is introduced later.

→ Fix: Safer to pass `path` as an argument and slice/push locally. If you want to keep it for perf, fine, but document it and don’t loosen discipline.

---

### ❌ `ignore` is called even when `deep` is false

* The `ignore` call should only happen in deep mode.
* Currently it happens even in non-deep array mode and root `_sortKeys`.

→ Fix: Wrap `ignore` logic in `if (deep && ignore)` or clarify in docs that ignore applies even when `deep: false`.

---

### ❌ `compare` is used even when not sorting deeply

* `_sortKeys` is used for top-level object even when `deep: false`, so `compare` always applies. That’s expected behavior, but ensure test coverage matches it.

→ Optional: No fix needed unless behavior is wrong.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
You need to update the readme too.

File: readme.md [hunk 1]
```diff
@@ -59,3 +59,9 @@ Recursively sort keys, including keys of objects inside arrays.
 Type: `Function`
 
 [Compare function.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort)
+
+##### ignore
+
+Type: `Function`
+
+Ignore fields that meet certain conditions to avoid sorting.
```
Comment (sindresorhus) [inline readme.md:67]: Missing a lot of info.

File: test.js [hunk 2]
```diff
@@ -152,3 +152,44 @@ test('keeps property descriptors intact', t => {
...
+	};
+	const sorted = sortKeys(object, {
+		ignore: ({key}) => key === 'e',
+		deep: true,
+	});
+	t.deepEqual(Object.keys(sorted), ['a', 'b', 'c']);
+	t.deepEqual(Object.keys(sorted.c), ['e', 'f']);
+	t.deepEqual(Object.keys(sorted.c.f), ['a', 'd']);
+	t.deepEqual(Object.keys(sorted.c.e), ['h', 'g']);
+});
```
Comment (sindresorhus) [inline test.js:195]: Need more tests to check edge-cases. Feel free to ask AI to generate those.

File: test.js [hunk 3]
```diff
@@ -152,3 +152,44 @@ test('keeps property descriptors intact', t => {
...
+				a: 0,
+			},
+			b: 0,
+		},
+	};
+	const sorted = sortKeys(object, {ignore: ({depth}) => depth >= 2, deep: true});
+	t.deepEqual(Object.keys(sorted), ['b', 'c']);
+	t.deepEqual(Object.keys(sorted.c), ['b', 'd']);
+	t.deepEqual(Object.keys(sorted.c.d), ['e', 'a']);
+});
```
Comment (sindresorhus) [inline test.js:171]:
```suggestion
});

```

File: index.d.ts [hunk 4]
```diff
@@ -1,15 +1,32 @@
...
+};
+
 export type Options = {
 	/**
 	Recursively sort keys, including keys of objects inside arrays.
 
 	@default false
 	*/
-	readonly deep?: boolean;
+	readonly deep?: boolean | number;
```
Comment (sindresorhus) [inline index.d.ts:19]: ⬆️

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by btea** [comment]
Thank you for the review. I have updated it.

**Comment by sindresorhus** [comment]
AI review:

___

One actionable bug
- Accessor properties when not ignored can still throw:
  - In the normal (non-ignored) path you always do:
    - Read `value = object[key]` (invokes the getter), possibly recurse, then
    - `Object.defineProperty(result, key, {...descriptor, value: newValue});`
  - If the original property is an accessor (has get/set), mixing it with a valu
e produces an invalid descriptor and throws. There’s no test covering this path
right now, which is why it slips through.
  - Minimal fix in the non-ignored path: detect accessor descriptors and copy th
em as-is instead of writing a value.

Suggested patch (index.js)
```diff
@@
-    if (deep && Array.isArray(value)) {
-      newValue = deepSortArray(value, path);
-    } else {
-      newValue = deep && isPlainObject(value) ? _sortKeys(value, path) : value;
-    }
-
-    path.pop();
-    Object.defineProperty(result, key, {
-      ...Object.getOwnPropertyDescriptor(object, key),
-      value: newValue,
-    });
+    if (deep && Array.isArray(value)) {
+      newValue = deepSortArray(value, path);
+    } else {
+      newValue = deep && isPlainObject(value) ? _sortKeys(value, path) : value;
+    }
+
+    path.pop();
+    const descriptor = Object.getOwnPropertyDescriptor(object, key);
+    if (descriptor && ('get' in descriptor || 'set' in descriptor)) {
+      // Preserve accessor descriptor as-is
+      Object.defineProperty(result, key, descriptor);
+    } else {
+      Object.defineProperty(result, key, {
+        ...descriptor,
+        value: newValue,
+      });
+    }
```

Optional refinements
- Use Object.hasOwn in the accessor checks for clarity:
  - `Object.hasOwn(descriptor, 'get') || Object.hasOwn(descriptor, 'set')`
- Micro-nit: consider renaming `originKeys` → `originalKeys`.

Docs
- Sentinel: Please document that the callback may receive `key: undefined` to de
cide whether to sort the current object’s own keys (object-level decision). Righ
t now the README says `key` is `string | number`, which conflicts with both the
implementation and typings.
- Depth semantics:
  - Per-key callback: `depth === path.length` (after pushing the key/index).
  - Object-level decision (key is undefined): `depth === path.length + 1`.
- Wording tweak: “skip sorting of certain options” → “skip sorting of certain ke
ys or branches.”
- Clarify behavior: If the object-level sentinel returns true, the current objec
t’s own key order is preserved, but children are still deep-sorted unless those
are also ignored.

Suggested README additions
```md
- `key` — `string | number | undefined`
  - When `undefined`, the callback decides whether to sort the current object’s
own keys (object-level decision).

Examples:

// Preserve top-level object key order, but still deep-sort children
sortKeys(input, {
  deep: true,
  ignore: ({key, path}) => key === undefined && path.length === 0
});

// Skip sorting at depth 3 only (object-level)
sortKeys(input, {
  deep: true,
  ignore: ({key, depth}) => key === undefined && depth === 3
});
```

Tests
- Please add a test to lock down the non-ignored accessor behavior (after applyi
ng the code fix above):
```js
test('accessor property is preserved when not ignored', t => {
  const input = {};
  Object.defineProperty(input, 'foo', {
    get() { return 123; },
    enumerable: true,
    configurable: true
  });
  const output = sortKeys(input, {deep: true});
  const desc = Object.getOwnPropertyDescriptor(output, 'foo');
  t.truthy(desc);
  t.is(typeof desc.get, 'function');
  t.false('value' in desc);
  t.is(output.foo, 123);
});
```

**Comment by btea** [comment]
image

Using `Object.hasOwn` will result in the error shown in the figure, so it is not modified.
